File: frontend/netlify/functions/lib/openrouter.js
/**
 * OpenRouter API client for making LLM requests.
 */

const OPENROUTER_API_URL = 'https://openrouter.ai/api/v1/chat/completions';

/**
 * Query a single model via OpenRouter API.
 * @param {string} model - OpenRouter model identifier
 * @param {Array} messages - Array of message objects with role and content
 * @param {number} timeout - Request timeout in milliseconds
 * @returns {Object|null} Response with content, or null if failed
 */
export async function queryModel(model, messages, timeout = 120000) {
  const apiKey = process.env.OPENROUTER_API_KEY;
  if (!apiKey) {
    console.error('OPENROUTER_API_KEY not set');
    return null;
  }

  const controller = new AbortController();
  const timeoutId = setTimeout(() => controller.abort(), timeout);

  try {
    const response = await fetch(OPENROUTER_API_URL, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ model, messages }),
      signal: controller.signal,
    });

    clearTimeout(timeoutId);

    if (!response.ok) {
      const errorText = await response.text();
      console.error(`OpenRouter API error for ${model}: ${response.status} ${errorText}`);
      return null;
    }

    const data = await response.json();
    const message = data.choices?.[0]?.message;

    return {
      content: message?.content || '',
      reasoning_details: message?.reasoning_details,
    };
  } catch (error) {
    clearTimeout(timeoutId);
    console.error(`Error querying model ${model}:`, error.message);
    return null;
  }
}

/**
 * Query multiple models in parallel.
 * @param {Array<string>} models - Array of model identifiers
 * @param {Array} messages - Messages to send to each model
 * @returns {Object} Map of model identifier to response
 */
export async function queryModelsParallel(models, messages) {
  const tasks = models.map((model) => queryModel(model, messages));
  const responses = await Promise.all(tasks);

  const result = {};
  models.forEach((model, index) => {
    result[model] = responses[index];
  });

  return result;
}

File: backend/openrouter.py
"""OpenRouter API client for making LLM requests."""

import asyncio
from typing import Any, Dict, List, Optional

import httpx

from .config import OPENROUTER_API_KEY, OPENROUTER_API_URL


async def query_model(
    model: str,
    messages: List[Dict[str, str]],
    timeout: float = 120.0,
) -> Optional[Dict[str, Any]]:
    """
    Query a single model via OpenRouter.

    Returns dict with 'content' and optional 'reasoning_details', or None on failure.
    """
    headers = {
        "Authorization": f"Bearer {OPENROUTER_API_KEY}",
        "Content-Type": "application/json",
    }
    payload = {"model": model, "messages": messages}

    try:
        async with httpx.AsyncClient(timeout=timeout) as client:
            response = await client.post(OPENROUTER_API_URL, headers=headers, json=payload)
            response.raise_for_status()

            message = response.json()["choices"][0]["message"]
            return {
                "content": message.get("content"),
                "reasoning_details": message.get("reasoning_details"),
            }
    except Exception as e:
        print(f"Error querying model {model}: {e}")
        return None


async def query_models_parallel(
    models: List[str],
    messages: List[Dict[str, str]],
) -> Dict[str, Optional[Dict[str, Any]]]:
    """Query multiple models in parallel. Returns dict mapping model ID to response."""
    tasks = [query_model(model, messages) for model in models]
    responses = await asyncio.gather(*tasks)
    return dict(zip(models, responses))
